### PR TITLE
HPCC-26443 Allow query vs other (third party) memory requirements

### PR DESCRIPTION
**Pull request by jakesmith**
Containerized only.

In order to explictly define how much of the resourced memory should be
dedicated to Thor's internal memory (roxiemem) vs other uses within the
same container, e.g. by 3rd parties, add a memory section which can
define memory characteristics.

Also allow these settings to be defined per workunit.

May expand at some point to also define things like amount to use for
different caching sizes etc.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: helm/hpcc/values.schema.json [hunk 1]
```diff
@@ -1192,6 +1205,12 @@
           "type": "object",
           "additionalProperties": { "type": "string" }
         },
+        "workerMemory": {
```
Comment (richardkchapman) [inline helm/hpcc/values.schema.json:1208]: Should this also be applied to Roxie, hthor?

File: thorlcr/graph/thgraph.hpp [hunk 2]
```diff
@@ -69,6 +69,23 @@
...
+ *
+ * For historical reasons the default in bare-metal has always been a
+ * conservative reserve of 25%.
+ *
+ * NB: These percentages do not apply if the memory amount has been configured
+ * manually via 'globalMemorySize' and 'masterMemorySize'
+ */
+
+#ifdef _CONTAINERIZED
+constexpr float roxieMemPercentage = 10.0;
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply thorlcr/graph/thgraph.hpp:83]: It is not clear from the comment above whether this is memory reserved by this process, or memory reserved for other items that is not used by this process.  It took me a while to work it out.

File: helm/hpcc/values.schema.json [hunk 3]
```diff
@@ -473,6 +473,19 @@
     "resources": {
       "type": "object"
     },
+    "memory": {
```
Comment (ghalliday) [inline helm/hpcc/values.schema.json:476]: The possible settings should be documented in the jira, together with how they are processed.  (It was very hard to review the code without it.)

**Review by ghalliday — COMMENTED** [review]
@jakesmith a few comments.
I found it hard to follow the logic, probably because there is a mixture of how much not to allocate (with reserve), and how much to allocate (with new settings and other settings).  It might be clearer to use a percentage of memory used rather than reserved, but needs clear documentation in the jira.

**Review by jakesmith — COMMENTED** [review]
@ghalliday - please see replies and new commit.

**Comment by jakesmith** [comment]
>@ghalliday - please see replies and new commit.

@ghalliday - looks like I forgot to refresh the github review status.

File: thorlcr/graph/thgraphslave.cpp [hunk 4]
```diff
@@ -1686,7 +1686,48 @@ CJobSlave::CJobSlave(ISlaveWatchdog *_watchdog, IPropertyTree *_workUnitInfo, co
         pluginMap->loadFromList(pluginsList.str());
     }
     tmpHandler.setown(createTempHandler(true));
-    sharedAllocator.setown(::createThorAllocator(globalMemoryMB, sharedMemoryMB, numChannels, memorySpillAtPercentage, *logctx, crcChecking, usePackedAllocator));
+
+    /*
+     * Calculate maximum recommeded memory for this worker.
```
Comment (ghalliday) [inline thorlcr/graph/thgraphslave.cpp:1691]: trivial: recomme**n**ded

File: thorlcr/graph/thgraphslave.cpp [hunk 5]
```diff
@@ -1686,7 +1686,48 @@ CJobSlave::CJobSlave(ISlaveWatchdog *_watchdog, IPropertyTree *_workUnitInfo, co
...
+
+    // @localThor mode - 25% is used for manager and 50% is used for workers
+    if (globals->getPropBool("@localThor") && (0 == globals->getPropInt("@masterMemorySize")))
+    {
+        /* In this mode, 25% is reserved for manager,
+         * 50% for the workers.
+         * Meaning this workers' recommendedMaxPercentage is remaining percentage */
+        recommendedMaxPercentage -= 25.0; // for manager
+        float pctPerWorker = 50.0 / numWorkersPerNode;
+        recommendedMaxPercentage -= pctPerWorker * (numWorkersPerNode-1);
```
Comment (ghalliday) [inline thorlcr/graph/thgraphslave.cpp:1715]: minor: I think you can say

recommendedMaxPercentage = pctPerWorker;

File: thorlcr/graph/thgraphslave.cpp [hunk 6]
```diff
@@ -1686,7 +1686,48 @@ CJobSlave::CJobSlave(ISlaveWatchdog *_watchdog, IPropertyTree *_workUnitInfo, co
...
+    }
+    else
+    {
+        // deduct percentage for all other workers from max percentage
+        float pctPerWorker = defaultPctSysMemForRoxie / numWorkersPerNode;
+        recommendedMaxPercentage -= pctPerWorker * (numWorkersPerNode-1);
```
Comment (ghalliday) [inline thorlcr/graph/thgraphslave.cpp:1721]: similarly
`recommendedMaxPercentage = pctPerWorker;`

**Review by ghalliday — COMMENTED** [review]
@jakesmith a few minor comments.

File: helm/hpcc/values.schema.json [hunk 7]
```diff
@@ -473,6 +473,20 @@
...
+      "type": "object",
+      "properties": {
+        "query": {
+          "type": "string",
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply helm/hpcc/values.schema.json:480]: no a string because it's a user friendly format string, as with the resources section, e.g. "3Mi"

File: thorlcr/graph/thgraphslave.cpp (hunk 5 above)
Comment (jakesmith) [reply thorlcr/graph/thgraphslave.cpp:1715]: yep, you're right, changed.

File: thorlcr/graph/thgraphslave.cpp (hunk 6 above)
Comment (jakesmith) [reply thorlcr/graph/thgraphslave.cpp]: changed

File: thorlcr/graph/thgraphslave.cpp (hunk 4 above)
Comment (jakesmith) [reply thorlcr/graph/thgraphslave.cpp]: changed

**Review by ghalliday — APPROVED** [review]

File: helm/hpcc/values.schema.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (jakesmith) [reply helm/hpcc/values.schema.json:480]: I can't see "stringOrInt" as a json schema type (http://json-schema.org/understanding-json-schema/reference/type.html)
or did you mean "oneOf" and allow "string" or "number" ? 
Not sure point though of adding "int" if you can add anything via "string" ?
